xarray/tests/test_dask.py

```diff
@@ -878,3 +878,77 @@ def test_dask_layers_and_dependencies():
     assert set(x.foo.__dask_graph__().dependencies).issuperset(
         ds.__dask_graph__().dependencies
     )
+
+
+def make_da():
+    return xr.DataArray(
+        np.ones((10, 20)),
+        dims=["x", "y"],
+        coords={"x": np.arange(10), "y": np.arange(100, 120)},
+        name="a",
+    ).chunk({"x": 4, "y": 5})
+
+
+def make_ds():
+    map_ds = xr.Dataset()
+    map_ds["a"] = map_da
+    map_ds["b"] = map_ds.a + 50
+    map_ds["c"] = map_ds.x + 20
+    map_ds = map_ds.chunk({"x": 4, "y": 5})
+    map_ds["d"] = ("z", [1, 1, 1, 1])
+    map_ds["z"] = [0, 1, 2, 3]
+    map_ds["e"] = map_ds.x + map_ds.y + map_ds.z
+    map_ds.attrs["test"] = "test"
+
+    return map_ds
+
+
+# work around mypy error
+# xarray/tests/test_dask.py:888: error: Dict entry 0 has incompatible type "str": "int"; expected "Hashable": "Union[None, Number, Tuple[Number, ...]]"
+map_da = make_da()
+map_ds = make_ds()
+
+
+def simple_func(obj):
+    result = obj.x + 5 * obj.y
+    return result
+
+
+def complicated_func(obj):
+    new = obj.copy()
+    new = (
+        new[["a", "b"]]
+        .rename({"a": "new_var1"})
+        .expand_dims(k=[0, 1, 2])
+        .transpose("k", "y", "x")
+    )
+    new["b"] = new.b.astype("int32")
+    return new
+
+
+def test_map_blocks_error():
+    def bad_func(darray):
+        return (darray * darray.x + 5 * darray.y)[:1, :1]
+
+    with raises_regex(ValueError, "Length of the.* has changed."):
+        xr.map_blocks(bad_func, map_da).compute()
+
+
+@pytest.mark.parametrize(
+    "func, obj",
+    [[simple_func, map_da], [simple_func, map_ds], [complicated_func, map_ds]],
+)
+def test_map_blocks(func, obj):
+
+    actual = xr.map_blocks(func, obj)
+    expected = func(obj)
+    xr.testing.assert_equal(expected, actual)
+
+
+@pytest.mark.parametrize("obj", [map_da, map_ds])
+def test_map_blocks_args(obj):
+    import operator
+
+    expected = obj + 10
+    actual = xr.map_blocks(operator.add, obj, 10)
+    xr.testing.assert_equal(expected, actual)
```

Comment:
Missing use cases: 
- func(DataArray) -> Dataset 
- func(Dataset) -> DataArray
- func output contains less variables
- func output contains new variables
- func changes dtypes
- *args, **kwargs are passed through
- IndexVariables don't accidentally cause the whole graph to be computed (the logic you wrote in the main function is quite subtle!)

Reply:
Could you add convenience methods?
```python
class Dataset:
    ...
    def map_blocks(self, func, *args, **kwargs):
        from .parallel import map_blocks
        return map_blocks(func, self, *args, **kwargs)
```

Reply:
I'd like to but i was running into circular import issues when i tried this a while back. Any thoughts on fixing that?

Reply:
I wrote the whole thing above...

Reply:
I agree that these are interesting use cases, but let's not block this PR on all this functionality. For now, I think it would be useful even if this function only works on function that accept/return objects with all the same metadata.

For completeness, we could also add one more use-case:
- func output contains less (or more) dimensions

Reply:
> I wrote the whole thing above...

=) sorry! didn't look too closely.

Reply:
For

> IndexVariables don't accidentally cause the whole graph to be computed (the logic you wrote in the main function is quite subtle!)

take a look at https://stackoverflow.com/questions/53289286/determine-how-many-times-dask-computed-something

Reply:
Yeah we have `CountingScheduler` in `test_dask.py` that does what we want.